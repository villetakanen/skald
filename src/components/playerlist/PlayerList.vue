<template>
  <div>
    <v-data-table
      :items="players.list"
      :loading="players.loading"
      :headers="headers">
      <template v-slot:item.actions="{ item }">
      <v-icon
        small
        class="mr-2"
        @click="editItem(item)"
      >
        mdi-pencil
      </v-icon>
      <v-icon
        small
        @click="removePlayer(item)"
      >
        mdi-delete
      </v-icon>
    </template>
    </v-data-table>
  </div>
</template>

<script lang="ts">
import firebase, { FirebaseError } from 'firebase/app'
import 'firebase/firestore'
import Vue from 'vue'
import VueCompositionApi, { defineComponent } from '@vue/composition-api'
import { usePlayers, Player } from '@/lib/usePlayers'
import { useSite } from '@/lib/useSite'
import { useAppState } from '@/lib/useAppState'
Vue.use(VueCompositionApi)

export default defineComponent({
  setup (props) {
    const { players } = usePlayers()
    const { site } = useSite()
    const { raiseError } = useAppState()
    const headers = [
      { text: 'Nick', value: 'nick' },
      { text: 'Tags', value: 'tags' },
      { text: 'Actions', value: 'actions' }
    ]
    const removePlayer = (player:Player):void => {
      const db = firebase.firestore()
      const playerRef = db.collection('sites').doc(site.value.siteid).collection('players').doc(player.uid)
      playerRef.delete().catch((error:Error) => {
        raiseError(error as FirebaseError)
      })
    }
    return { players, headers, removePlayer }
  }
})
</script>
