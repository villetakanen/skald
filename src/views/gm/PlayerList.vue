<template>
  <v-container>
    <v-row>
      <v-col>
        <TabTitle
          :sub="$t('view_gm_players.title')"
          :topic="siteid"/>
      </v-col>
    </v-row>
    <v-row>
      <v-col>
        <Loading v-if='loading'/>
        <v-card v-if="!loading">
          <v-card-title>{{$t('view_gm_players.title')}}</v-card-title>
          <v-card-text>
            <AddPlayersAction :siteid="siteid"/>
            <template v-for="(player, index) in players">
              <div v-bind:key="index">
                <div class="playerlist-cell">
                  {{player.nick}}
                </div>
                <div class="playerlist-cell">
                  <template v-for="(tag, tagIndex) in player.tags">
                    <span v-bind:key="tagIndex" class="playerlist-tag">{{tag}}</span>
                  </template>
                </div>
              </div>
            </template>
          </v-card-text>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue'
import VueCompositionApi, { /* reactive, toRefs, */ ref, defineComponent, onMounted, onUnmounted } from '@vue/composition-api'
import Loading from '../../components/Loading.vue'
import TabTitle from '../../components/TabTitle.vue'
import AddPlayersAction from '../../components/site/AddPlayersAction.vue'
import firebase from 'firebase/app'
import 'firebase/firestore'

Vue.use(VueCompositionApi)

export default defineComponent({
  components: {
    Loading,
    TabTitle,
    AddPlayersAction
  },
  props: {
    siteid: {
      type: [String],
      required: true
    }
  },
  setup (props) {
    const loading = ref(true)
    const pr:object[] = []
    const players = ref(pr)
    let unsubscribePlayers = () => {}
    onMounted(() => {
      const db = firebase.firestore()
      const playerRef = db.collection('sites').doc(props.siteid).collection('players')
      unsubscribePlayers = playerRef.onSnapshot((data) => {
        const playerData:object[] = []
        data.forEach((player) => {
          playerData.push(player.data())
        })
        players.value = playerData
        loading.value = false
      })
    })
    onUnmounted(() => { unsubscribePlayers() })
    return { loading, players }
  }
})
</script>

<style lang="scss" scoped>
.playerlist-cell {
  display: inline-block;
  padding-left: 8px;
}
.playerlist-tag {
  background-color: RGBA(170, 170, 170, 0.5);
  border-radius: 4px;
  padding: 0px 4px;
  display: inline-block;
  margin-left: 4px;
}

</style>
