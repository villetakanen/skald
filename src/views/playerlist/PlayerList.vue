<template>
  <v-container>
    <v-row>
      <v-col>
        <v-card>
          <v-card-text>
            <PlayerList/>
          </v-card-text>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import Vue from 'vue'
import VueCompositionApi, { /* reactive, toRefs, */ ref, defineComponent, onMounted, onUnmounted } from '@vue/composition-api'
import PlayerList from '@/components/playerlist/PlayerList.vue'
Vue.use(VueCompositionApi)

export default defineComponent({
  components: {
    PlayerList
  }
})
</script>
