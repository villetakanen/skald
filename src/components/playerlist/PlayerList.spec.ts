import { shallowMount } from '@vue/test-utils'
import PlayerList from './PlayerList.vue'

describe('Mounted App', () => {
  it('Renders with usePlayers init for "mekanismi"', () => {
    const wrapper = shallowMount(PlayerList)
    expect(wrapper.element).toMatchSnapshot()
  })
})
