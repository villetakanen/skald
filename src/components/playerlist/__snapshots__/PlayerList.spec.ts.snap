// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted App Renders with usePlayers init for "mekanismi" 1`] = `<div />`;
